src/wat-writer.cc

```diff
@@ -807,9 +871,46 @@ void WatWriter::PushExpr(const Expr* expr,
     expr_tree_stack_.emplace_back(expr);
     FlushExprTreeStack();
   }
+  if (s_trace) {
+    fprintf(stderr, "-> PushExpr(%s, %" PRIindex ", %" PRIindex")\n",
+            GetExprTypeName(*expr), operand_count, result_count);
+  }
+}
+
+void WatWriter::PushCatch(const Catch* catch_) {
+  if (s_trace)
+    fprintf(stderr, "-> PushCatch()\n");
+  ExprTree tree(catch_);
+  expr_tree_stack_.push_back(std::move(tree));
+  FlushExprTreeStack();
+  if (s_trace)
+    fprintf(stderr, "<- PushCatch()\n");
+}
+
+void WatWriter::FlushCatchTree(const ExprTree& expr_tree) {
```

Comment:
Now you know why the tracing. It was not very obvious to me (until I had some tracing), what was even going on. The code of this PR is quite different than initially written, and only got simplified once I saw the picture of what the code was doing. Will simplify.